### DISCUSSION TITLE
UDP routing issue

### DISCUSSION THREAD

**Discussion by pljones** [jamulussoftware/jamulus#2811] [Technical Support & Help] [answered]
Hi all,

I've raised this in the past but always dropped it...

![image](https://user-images.githubusercontent.com/1549463/187063801-405247d1-ba61-4276-9ae4-d5a2f70a78e7.png)

* The image on the left is `mtr -f 2 93.42.88.44` - that is, it's a progressive ICMP ping trace route.
* The image on the right is `mtr -f 2 -u -L 55859 -P 22124 93.42.88.44` - that is, much closer to the Jamulus protocol.

The IP and target port chosen here are visible and work on Jamulus Explorer
![image](https://user-images.githubusercontent.com/1549463/187063892-10ba8f04-0605-4c7d-81b1-fdc548894cb6.png)
and the source port is open at my end.  It's not just that host - if I run with `showallservers`, there's a lot of servers I see with empty ping:
![image](https://user-images.githubusercontent.com/1549463/187064259-52137256-0f03-418b-9cf4-718b2b79db24.png).

There's clearly some issue between `ip-plus.net` and the target network.  So what I'm looking for here are suggestions on how to work around that.

Any help gratefully received.

(Obviously, _some_ servers are set up wrongly - that's clear on Jamulus Explorer.  I'm not worried about those.  However, the Rainbow boxes appear to be good, as do many of the others that don't return UDP pings.)

-- Peter

**Comment by ann0see** [comment]
Probably it would be worth contacting the people running servers which show this issue?

**Reply by pljones** [reply]
What, the ones not returning any details? `???` doesn't give me a lot to go on.

**Reply by ghost** [reply]
I think the ??? routers are not passing UDP traffic (or blocking the UDP ports you used - try some well known UDP ports). Find out who they are and let them know. Note that the i00rom-005-ae0.bb.ip-plus.net router knows the ipaddr of the first ??? router.

**Reply by ann0see** [reply]
As it seems to be a network related issue, and you know the IP and server name from explorer.jamulus.io, the most straight forward way to debug it, is to see "the other side". Rainbow's Welcome message even shows an E-Mail address.

Then you can compare settings and diagnose from both sides.

**Reply by pljones** [reply]
I don't know what you mean.  I can ICMP ping the final host.  A UDP "ping" to the port reported as open on Jamulus Explorer fails.  The last ICMP host reporting before the `???` entries is the same as the lsat working host before the UDP failure.

Can you give me an explicit command to try?

**Reply by ann0see** [reply]
I mean that you should run traceroute on the host 93.42.88.44 to your server.

I suspect - since I can't connect directly via `jamulus -c 93.42.88.44` that they block inbound traffic and connections only work via the directory server magic

**Reply by pljones** [reply]
> I mean that you should run traceroute on the host 93.42.88.44 to your server.

Sorry, I don't understand that.  I have no access to 93.42.88.44 and I'm trying to access it as a normal Jamulus client.

**Reply by ann0see** [reply]
Yes. Therefore, 
>  it would be worth contacting the people running servers which show this issue

There seems to be an E-Mail in the server welcome message: 
![Bildschirmfoto](https://user-images.githubusercontent.com/20726856/187093499-528dcc68-6c90-4041-bc91-5543e18326b1.png)

I'd ask for their firewall/networking settings.

**Reply by pljones** [reply]
I'm almost 100% certain it's not a problem with their server.  Jamulus Explorer connects just fine, as do other clients.  And it's not specific to this server, either - I'd have to do the same for about 50.  So I don't think that's a viable approach.

**Reply by ann0see** [reply]
I mean, I can't connect directly. Only if I "requested access" via the GUI and the Directory Server hole punching a Connection works. So I do suspect that it's a thing on their side.

**Reply by pljones** [reply]
> I mean, I can't connect directly. Only if I "requested access" via the GUI and the Directory Server hole punching a Connection works. So I do suspect that it's a thing on their side.

Ahhh...  Now that's interesting.  I've never understood how the directory hole punch works... but my client will have the same IP address as the directory... just a different source port.

**Reply by pljones** [reply]
Ahh... right, so the "hole punch" says:
```
pConnLessProtocol->CreateCLSendEmptyMesMes ( siCurListEntry.HostAddr, InetAddr );
```
where `siCurListEntry.HostAddr` is the server address and `InetAddr` is the client address (requesting the server list).

So _if and only if_ the server _requires_ the hole punch, it _will fail_ if `InetAddr` is on the same LAN as the directory.  That's because the server will get told the LAN address of the client, which is useless!!

That does rather explain it.

Mmm...  I think I might be able to fix that...

**Reply by pljones** [reply]
Nope, sadly my idea (i.e. "do what the directory server does - return zero") didn't work for the "fake client ping".

**Reply by pljones** [reply]
Of course, if the directory has --serverpublicip set, then it might work...  Hm...  I'll keep thinking about it.

**Comment by hoffie** [comment]
Maybe so-called Looking Glasses by several ISPs could help narrowing it down from several sides. I'm not sure though if you get to choose the UDP port there though.
https://looking.house/index.php seems to collect some of those instances, but I don't see a port option there. :(

**Comment by pljones** [comment, marked as answer]
I've raised #2818 to cover what I think the issue is - thanks to @ann0see for the poke in the right direction.  I've a PR on its way with a workaround.

**Reply by ann0see** [reply]
I hope it's that.
